fem/nonlinearform.hpp

```diff
@@ -76,7 +78,22 @@ class NonlinearForm : public Operator
    { }
 
    /// Set the desired assembly level. The default is AssemblyLevel::NONE.
-   /** This method must be called before assembly. */
+   /** For nonlinear operators, the "matrix" assembly levels usually do not make
```

Comment:
I'm not sure what you mean by EA in this context. Typically, EA means we assemble an element matrix for each mesh element which represents the local element action of the operator, so this only makes sense for linear operators. Do you have something different in mind for representing nonlinear operators on element level?

Reply:
In my application, we have the `NonlinearForm::GetGradient` return the operator of the Jacobian of our PDE to be used in a NR solver. The Jacobian itself is nontrivial to compute, so we already have stored as either a 6x6xnqpts array per element for EA/FA or the more general 3x3x3x3xnqpts per element for the PA application. Then our `Nonlinear::Mult` just returns the divergence of our stress (also a nonlinear function). Now, our Jacobian and stress are both calculated at the same time which is why we went with a `NonlinearForm` to represent things rather than trying to things represented as something else.

Our PA within a `NonlinearFormIntegrator` for the gradient can be represented by Eq. 6 in this paper: https://doi.org/10.1002/nme.1620050108

Our EA within a `NonlinearFormIntegrator` for the gradient can be represented by Eq. 4 in this paper: https://doi.org/10.1002/nme.1620190910

While the papers are looking at the trivial elastic case, the general formulation works just as well in the nonlinear case.

Reply:
Generally, there are two separate assembly-levels associated with a nonlinear operator -- one for the operator action (which can be NONE or PA) and another for its gradient (which can be NONE, PA, EA, or FA).

The doxygen comment here was meant to explain what the current behavior is when you set the assembly-level -- this is the assembly level for the operator action. The assembly-level for the gradient cannot be set at the moment (this can be added in the future) and currently it is either FA (when the assembly-level for the operator action is NONE -- this is the "standard" behavior with no `NonlinearFormExtension`) or PA (when the assembly-level for the operator action is PA). In the future we can definitely support combinations like NONE+NONE, PA+EA, PA+FA, etc, as we add the respective implementations.

Reply:
Okay, I definitely missed the distinction for that. Sorry about the confusion. Since that's the case, I'm not aware of any need for anything other than the 2 you listed for the operator.